fog/view/connection/src/fog_view_router_client.rs

```diff
@@ -2,68 +2,230 @@
+            .request()
+            .expect("Could not retrieve grpc sender and receiver.");
 
         Self {
-            fog_view_router_client,
             logger,
+            attest_cipher: None,
+            _fog_view_router_client: fog_view_router_client,
+            request_sender,
+            response_receiver,
+            uri,
+            verifier,
         }
     }
 
-    /// Makes streaming requests to the fog view router service.
-    pub async fn request(&self) -> Result<(), grpcio::Error> {
-        let (mut sink, mut receiver) = self.fog_view_router_client.request()?;
-        let attested_message = Message::new();
+    fn is_attested(&self) -> bool {
+        self.attest_cipher.is_some()
+    }
+
+    async fn attest(&mut self) -> Result<VerificationReport, Error> {
+        // If we have an existing attestation, nuke it.
+        self.deattest();
+
+        let mut csprng = McRng::default();
+
+        let initiator = Start::new(
+            self.uri
+                .responder_id()
+                .map_err(|_| Error::Attestation())?
+                .to_string(),
+        );
+
+        let init_input = ClientInitiate::<X25519, Aes256Gcm, Sha512>::default();
+        let (initiator, auth_request_output) = initiator
+            .try_next(&mut csprng, init_input)
+            .map_err(|_| Error::Attestation())?;
+
+        let attested_message: AuthMessage = auth_request_output.into();
         let mut request = FogViewRouterRequest::new();
-        request.set_query(attested_message);
-        let send = async move {
-            for i in 0..5 {
-                log::info!(self.logger, "Sending message {}", i);
-                sink.send((request.clone(), grpcio::WriteFlags::default()))
-                    .await?;
-            }
-            sink.close().await?;
-            Ok(()) as Result<(), grpcio::Error>
+        request.set_auth(attested_message);
+        self.request_sender
+            .send((request.clone(), grpcio::WriteFlags::default()))
+            .await?;
+
+        let mut response = self.response_receiver.try_next().await?.unwrap();
+        let auth_response_msg = response.take_auth();
+
+        // Process server response, check if key exchange is successful
+        let auth_response_event =
+            AuthResponseInput::new(auth_response_msg.into(), self.verifier.clone());
+        let (initiator, verification_report) = initiator
+            .try_next(&mut csprng, auth_response_event)
+            .map_err(|_| Error::Attestation())?;
+
+        self.attest_cipher = Some(initiator);
+
+        Ok(verification_report)
+    }
+
+    fn deattest(&mut self) {
+        if self.is_attested() {
+            log::trace!(
+                self.logger,
+                "Tearing down existing attested connection and clearing cookies."
+            );
+            self.attest_cipher = None;
+        }
+    }
+
+    /// Makes streaming requests to the fog view router service.
+    pub async fn query(
```

Comment:
:thought_balloon: More food for thought than anything else.
This method is a bit large, I could be missing some constraints, but it feels like there are 3 more or less independent pieces:
1. Build the request
2. do the talking (send and wait)
3. Decode the response

I would vote to break up functions like this (going forward), but I know that can often be subjective

Reply:
This is a good idea, something that could be addressed in a follow up.